core/src/main/java/de/jplag/Submission.java

```diff
@@ -157,7 +136,7 @@ int getSimilarityDivisor(boolean subtractBaseCode) {
      * @return unmodifiable list of tokens generated by parsing the submission.
      */
     public List<Token> getTokenList() {
-        return tokenList == null ? null : Collections.unmodifiableList(tokenList);
+        return tokenList;
```

Comment:
I'd create the unmodifiable list here. In the future, you might miss the constraint in other methods.

Reply:
I moved it into parse() to avoid creating it on each call. This method is called quite extensively during GST, match merging, and token sequence normalization. However, I admit, this may be premature optimization. What is your take on that?